gdal/frmts/msg/msgdataset.cpp

```diff
@@ -345,31 +343,22 @@ MSGRasterBand::MSGRasterBand( MSGDataset *poDSIn, int nBandIn )
     // Find if we're dealing with MSG1, MSG2, MSG3 or MSG4
     // Doing this per band is the only way to guarantee time-series when the satellite is changed
 
-    std::string sPrologueFileName = poDSIn->command.sPrologueFileName(poDSIn->iCurrentSatellite, nBand);
+    int iCurrentSatellite = poDSIn->iCurrentSatellite; // Start with the satellite selected by the dataset
+
+    std::string sPrologueFileName = poDSIn->command.sPrologueFileName(iCurrentSatellite, 1);
     bool fPrologueExists = (access(sPrologueFileName.c_str(), 0) == 0);
 
-    // Make sure we're testing for MSG1,2,3 or 4 exactly once, start with the most recently used, and remember it in the static member for the next round.
-    if (!fPrologueExists)
+    // Make sure we're testing for MSG1,2,3 or 4 exactly once, start with the one of the dataset
+    int iTries = 1;
+    while (!fPrologueExists && (iTries < MAX_SATELLITES))
     {
-      poDSIn->iCurrentSatellite = 1 + poDSIn->iCurrentSatellite % MAX_SATELLITES;
-      sPrologueFileName = poDSIn->command.sPrologueFileName(poDSIn->iCurrentSatellite, nBand);
+      iCurrentSatellite = 1 + iCurrentSatellite % MAX_SATELLITES;
+      sPrologueFileName = poDSIn->command.sPrologueFileName(iCurrentSatellite, 1);
       fPrologueExists = (access(sPrologueFileName.c_str(), 0) == 0);
-      int iTries = 2;
-      while (!fPrologueExists && (iTries < MAX_SATELLITES))
-      {
-        poDSIn->iCurrentSatellite = 1 + poDSIn->iCurrentSatellite % MAX_SATELLITES;
-        sPrologueFileName = poDSIn->command.sPrologueFileName(poDSIn->iCurrentSatellite, nBand);
-        fPrologueExists = (access(sPrologueFileName.c_str(), 0) == 0);
-        ++iTries;
-      }
-      if (!fPrologueExists) // assume missing prologue file, keep original satellite number
-      {
-        poDSIn->iCurrentSatellite = 1 + poDSIn->iCurrentSatellite % MAX_SATELLITES;
-        sPrologueFileName = poDSIn->command.sPrologueFileName(poDSIn->iCurrentSatellite, nBand);
-      }
+      ++iTries;
     }
 
-    iSatellite = poDSIn->iCurrentSatellite; // From here on, the satellite that corresponds to this band is settled to the current satellite
+    iSatellite = iCurrentSatellite; // From here on, the satellite that corresponds to this band is settled to the current satellite
```

Comment:
Same for here, starting in 346. I guess before someone wanted to avoid race conditions (but it didn't work).